ADR/0058-maven-build-pipeline.md

```diff
@@ -0,0 +1,143 @@
+  catalogs.
+- The pipeline must invoke a reasonable default set of `mvn` commands (`mvn clean deploy`) to a
+  "dev" repository when code is merged. Code should not be deployed to a Maven repository for pull
+  request builds.
+- Developers must be able to provide arguments and CLI flags to the `mvn` command.
+- The build must produce OCI artifact references (`*_IMAGE_URL` and `*_IMAGE_DIGEST` or `IMAGES`)
+  for Tekton Chains to generate provenance and signatures.
+- Support for [multi-module](https://maven.apache.org/guides/mini/guide-multiple-modules.html)
+  projects is required.
+
+### Non-Goals
+
+- Support for [Maven 4](https://maven.apache.org/whatsnewinmaven4.html).
+- Support for Gradle.
+- Support for Kotlin, Scala, or other JVM-based programming languages.
+- Provision Maven repositories for Java projects.
+
+These can all be addressed through follow-up ADRs.
+
+## Decision
+
+The build pipeline will execute up to two "deployments":
+
+- The pipeline will first deploy Maven artifacts to a container registry using ORAS.
+- The pipeline will optionally execute a second deployment to a real Maven repository, if one is
+  configured.
+
+The core pipeline task, `maven-deploy-oras`, performs the following steps:
+
+1. **analyze:** Determine if deployment/snapshot repositories are configured and if the project
+   version is a `-SNAPSHOT`.
+2. **build-maven:** Execute the following to build the Maven project and stage the built artifacts
+   locally:
+   
+   ```
+   mvn clean ${MAVEN_EXTRA_GOALS} deploy ${MAVEN_EXTRA_OPTS} \
+     -DaltDeploymentRepository=local::file:./target/stage-deploy
+   ```
+
+3. **deploy-maven (Conditional):** If a real Maven repository is configured, perform a second
+   `mvn deploy` to the appropriate remote Maven repository.
+4. **Deploy-ORAS:** Use the ORAS CLI to push the full local staging directory tree to the container
+   registry:
+
+   ```
+   oras push ${IMG_URL} ./target/stage-deploy \
+     --artifact-type application/vnd.apache.maven.repository.v2.tar+gzip \
+     --export-manifest manifest-out.
+   ```
+
+   The artifact type `application/vnd.apache.maven.repository.v2.tar+gzip` communicates that this
+   image is non-runnable, with contents compressed with `gzip`. The output manifest is used for
+   Tekton Chains signing.
+
+The pipeline will also include the following tasks:
+
+- `generate-sbom`: this will generate the required component-level Software Bill of Materials from
+  multiple sources, aggregated with [Mobster](https://github.com/konflux-ci/mobster):
+  - The [CycloneDX Maven Plugin](https://github.com/CycloneDX/cyclonedx-maven-plugin) if the Maven
+    project has configured its use.
+  - [Syft](https://github.com/anchore/syft)
+  - Hermeto once Maven is supported as a package manager (design is
+    [under discussion](https://github.com/hermetoproject/hermeto/pull/1046))
+- `build-source-container`: this will build a source container, using existing tasks in the Konflux
```

Comment:
I don't know how useful a source container will be outside of Konflux. Instead, what if the build task also produces a sources JAR?

Reply:
I think we should pursue a more native approach for "packaging" source code.